File: effects.c
#include "effects.h"
#include "arena.h"
#include <stdint.h>
#include <string.h>

/* ---- int16 saturating clamp ---------------------------------- */

int16_t sat16(int32_t v) {
    if (v >  32767) return  32767;
    if (v < -32768) return -32768;
    return (int16_t)v;
}

/* ---- master-bus stereo delay --------------------------------- */
/* Two independent mono buffers, 250 ms long at 48 kHz. Standard
   feed-forward + feedback topology: out = dry + tap*wet; buffer-write
   = dry + tap*feedback. */
#define DELAY_SAMPLES  12000u

static int16_t *delay_l;
static int16_t *delay_r;
static uint32_t delay_idx      = 0;
static uint16_t delay_wet      = 100;  /* 0..256, mix amount */
static uint16_t delay_feedback = 140;  /* 0..200, capped to avoid runaway */

static int16_t *alloc_zero(uint32_t n_samples) {
    int16_t *p = arena_alloc(n_samples * sizeof(int16_t));
    memset(p, 0, n_samples * sizeof(int16_t));
    return p;
}

static void delay_init(void) {
    /* Idempotent: arena-allocate buffers on first call; subsequent
       calls just reset the index and clear the existing buffers. */
    if (!delay_l) {
        delay_l = alloc_zero(DELAY_SAMPLES);
        delay_r = alloc_zero(DELAY_SAMPLES);
    } else {
        memset(delay_l, 0, DELAY_SAMPLES * sizeof(int16_t));
        memset(delay_r, 0, DELAY_SAMPLES * sizeof(int16_t));
    }
    delay_idx = 0;
}

void delay_process(int16_t *buf, uint32_t frames) {
    for (uint32_t i = 0; i < frames; i++) {
        int32_t dry_l = buf[2 * i];
        int32_t dry_r = buf[2 * i + 1];
        int32_t tap_l = delay_l[delay_idx];
        int32_t tap_r = delay_r[delay_idx];

        int32_t out_l = dry_l + ((tap_l * (int32_t)delay_wet) >> 8);
        int32_t out_r = dry_r + ((tap_r * (int32_t)delay_wet) >> 8);

        int32_t fb_l = dry_l + ((tap_l * (int32_t)delay_feedback) >> 8);
        int32_t fb_r = dry_r + ((tap_r * (int32_t)delay_feedback) >> 8);

        delay_l[delay_idx] = sat16(fb_l);
        delay_r[delay_idx] = sat16(fb_r);

        if (++delay_idx >= DELAY_SAMPLES) delay_idx = 0;

        buf[2 * i]     = sat16(out_l);
        buf[2 * i + 1] = sat16(out_r);
    }
}

void delay_adjust_wet(int delta) {
    int v = (int)delay_wet + delta;
    if (v < 0)   v = 0;
    if (v > 256) v = 256;
    delay_wet = (uint16_t)v;
}

void delay_adjust_feedback(int delta) {
    int v = (int)delay_feedback + delta;
    if (v < 0)   v = 0;
    if (v > 200) v = 200;
    delay_feedback = (uint16_t)v;
}

uint16_t delay_get_wet(void)      { return delay_wet; }
uint16_t delay_get_feedback(void) { return delay_feedback; }

/* ---- Schroeder reverb ---------------------------------------- */
/* 4 parallel comb filters per channel, outputs summed and passed
   through 2 series all-pass filters per channel. Prime-number
   delays (Schroeder 1962, rescaled by 48000/44100) avoid metallic
   resonance. Slightly different L vs R delays preserve stereo
   separation in the tail. */
#define REV_C1L 1693
#define REV_C2L 1759
#define REV_C3L 1621
#define REV_C4L 1549
#define REV_C1R 1721
#define REV_C2R 1747
#define REV_C3R 1613
#define REV_C4R 1571
#define REV_AP1L 241
#define REV_AP2L 607
#define REV_AP1R 251
#define REV_AP2R 613
#define COMB_G   180             /* ~0.70 in 8.8 fixed, RT60 ~1.5 s */
#define AP_G     180

static int16_t *rev_c1l, *rev_c2l, *rev_c3l, *rev_c4l;
static int16_t *rev_c1r, *rev_c2r, *rev_c3r, *rev_c4r;
static int16_t *rev_ap1l, *rev_ap2l;
static int16_t *rev_ap1r, *rev_ap2r;
static uint16_t i_c1l, i_c2l, i_c3l, i_c4l;
static uint16_t i_c1r, i_c2r, i_c3r, i_c4r;
static uint16_t i_ap1l, i_ap2l, i_ap1r, i_ap2r;

static uint16_t reverb_wet      = 60;   /* 0..256, mix amount */
static int8_t   reverb_wet_bias = 0;    /* section-driven additive bias */

static void reverb_init(void) {
    /* Idempotent: arena-allocate comb/all-pass buffers on first call;
       subsequent calls clear the existing buffers and reset cursors. */
    if (!rev_c1l) {
        rev_c1l = alloc_zero(REV_C1L);  rev_c1r = alloc_zero(REV_C1R);
        rev_c2l = alloc_zero(REV_C2L);  rev_c2r = alloc_zero(REV_C2R);
        rev_c3l = alloc_zero(REV_C3L);  rev_c3r = alloc_zero(REV_C3R);
        rev_c4l = alloc_zero(REV_C4L);  rev_c4r = alloc_zero(REV_C4R);
        rev_ap1l = alloc_zero(REV_AP1L); rev_ap1r = alloc_zero(REV_AP1R);
        rev_ap2l = alloc_zero(REV_AP2L); rev_ap2r = alloc_zero(REV_AP2R);
    } else {
        memset(rev_c1l, 0, REV_C1L * 2); memset(rev_c1r, 0, REV_C1R * 2);
        memset(rev_c2l, 0, REV_C2L * 2); memset(rev_c2r, 0, REV_C2R * 2);
        memset(rev_c3l, 0, REV_C3L * 2); memset(rev_c3r, 0, REV_C3R * 2);
        memset(rev_c4l, 0, REV_C4L * 2); memset(rev_c4r, 0, REV_C4R * 2);
        memset(rev_ap1l, 0, REV_AP1L * 2); memset(rev_ap1r, 0, REV_AP1R * 2);
        memset(rev_ap2l, 0, REV_AP2L * 2); memset(rev_ap2r, 0, REV_AP2R * 2);
    }
    i_c1l = i_c2l = i_c3l = i_c4l = 0;
    i_c1r = i_c2r = i_c3r = i_c4r = 0;
    i_ap1l = i_ap2l = i_ap1r = i_ap2r = 0;
}

/* y[n] = x[n-D] + g*y[n-D] using delay-line as recirculating buffer. */
static inline int16_t comb_step(int16_t *buf, uint16_t size, uint16_t *idx, int16_t in) {
    int32_t tap = buf[*idx];
    int32_t w = in + ((tap * COMB_G) >> 8);
    buf[*idx] = sat16(w);
    *idx = (uint16_t)((*idx + 1) % size);
    return (int16_t)tap;
}

/* All-pass: same magnitude at every freq, shifts phases. Smooths
   the dense comb output into a continuous tail. */
static inline int16_t ap_step(int16_t *buf, uint16_t size, uint16_t *idx, int16_t in) {
    int32_t tap = buf[*idx];
    int32_t y = tap - ((in * AP_G) >> 8);
    int32_t w = in + ((tap * AP_G) >> 8);
    buf[*idx] = sat16(w);
    *idx = (uint16_t)((*idx + 1) % size);
    return sat16(y);
}

void reverb_process(int16_t *buf, uint32_t frames) {
    for (uint32_t i = 0; i < frames; i++) {
        int16_t in_l = buf[2 * i];
        int16_t in_r = buf[2 * i + 1];

        int32_t sum_l = (int32_t)comb_step(rev_c1l, REV_C1L, &i_c1l, in_l)
                      + comb_step(rev_c2l, REV_C2L, &i_c2l, in_l)
                      + comb_step(rev_c3l, REV_C3L, &i_c3l, in_l)
                      + comb_step(rev_c4l, REV_C4L, &i_c4l, in_l);
        sum_l >>= 2;
        int32_t sum_r = (int32_t)comb_step(rev_c1r, REV_C1R, &i_c1r, in_r)
                      + comb_step(rev_c2r, REV_C2R, &i_c2r, in_r)
                      + comb_step(rev_c3r, REV_C3R, &i_c3r, in_r)
                      + comb_step(rev_c4r, REV_C4R, &i_c4r, in_r);
        sum_r >>= 2;

        int16_t ap_l = ap_step(rev_ap1l, REV_AP1L, &i_ap1l, (int16_t)sum_l);
                ap_l = ap_step(rev_ap2l, REV_AP2L, &i_ap2l, ap_l);
        int16_t ap_r = ap_step(rev_ap1r, REV_AP1R, &i_ap1r, (int16_t)sum_r);
                ap_r = ap_step(rev_ap2r, REV_AP2R, &i_ap2r, ap_r);

        /* Apply section bias on top of user wet, clamped to [0, 256]. */
        int eff_wet = (int)reverb_wet + (int)reverb_wet_bias;
        if (eff_wet < 0)   eff_wet = 0;
        if (eff_wet > 256) eff_wet = 256;
        int32_t out_l = in_l + ((ap_l * eff_wet) >> 8);
        int32_t out_r = in_r + ((ap_r * eff_wet) >> 8);
        buf[2 * i]     = sat16(out_l);
        buf[2 * i + 1] = sat16(out_r);
    }
}

void reverb_adjust_wet(int delta) {
    int v = (int)reverb_wet + delta;
    if (v < 0)   v = 0;
    if (v > 256) v = 256;
    reverb_wet = (uint16_t)v;
}

uint16_t reverb_get_wet(void)               { return reverb_wet; }
void     reverb_set_wet_bias(int8_t bias)   { reverb_wet_bias = bias; }

/* ---- soft saturation ----------------------------------------- */
/* Cubic soft-clip: y = x - x^3 / 2^31. Linear for small x, smoothly
   compresses peaks. At full-scale int16 input (~32767) the output
   is ~50%; at typical levels (10-20% of full scale) the change is
   sub-1%. Adds gentle analog-tape warmth without affecting quiet
   signal character. */
static inline int16_t soft_sat(int16_t x) {
    int64_t x3 = (int64_t)x * x * x;
    int32_t cubic = (int32_t)(x3 >> 31);
    int32_t y = (int32_t)x - cubic;
    return sat16(y);
}

void saturate_process(int16_t *buf, uint32_t frames) {
    for (uint32_t i = 0; i < frames; i++) {
        buf[2 * i]     = soft_sat(buf[2 * i]);
        buf[2 * i + 1] = soft_sat(buf[2 * i + 1]);
    }
}

/* ---- master compressor + brickwall limiter ------------------- */
/* Feed-forward, stereo-linked. Envelope tracks max(|L|,|R|) so the
   gain reduction is identical on both channels and stereo imaging
   is preserved. 4:1 ratio above threshold, attack ~5 ms and release
   ~200 ms at 48 kHz, +1 dB makeup gain, brickwall ceiling just
   below int16 max so sat16 still has margin. */
#define COMP_THRESHOLD_DEFAULT  20000        /* int16 amplitude */
#define COMP_THRESHOLD_MIN       8000
#define COMP_THRESHOLD_MAX      30000
#define COMP_RATIO                  4        /* 4:1 above threshold */
#define COMP_ATTACK_COEF          268        /* (1-exp(-1/(0.005*48000)))*65536 */
#define COMP_RELEASE_COEF           7        /* (1-exp(-1/(0.200*48000)))*65536 */
#define COMP_MAKEUP_GAIN          288        /* ~+1 dB in 8.8 fixed */
#define LIMIT_CEILING           32000        /* brickwall, below int16 max */

static int32_t comp_env       = 0;
static int32_t comp_threshold = COMP_THRESHOLD_DEFAULT;

void compressor_process(int16_t *buf, uint32_t frames) {
    for (uint32_t i = 0; i < frames; i++) {
        int32_t l = buf[2 * i];
        int32_t r = buf[2 * i + 1];
        int32_t al = l < 0 ? -l : l;
        int32_t ar = r < 0 ? -r : r;
        int32_t a  = al > ar ? al : ar;

        /* Asymmetric envelope follower: fast attack, slow release. */
        if (a > comp_env) {
            comp_env += ((a - comp_env) * COMP_ATTACK_COEF) >> 16;
        } else {
            comp_env -= ((comp_env - a) * COMP_RELEASE_COEF) >> 16;
        }

        /* Gain reduction: 1.0 below threshold, compressed above.
           gain is 8.8 fixed; target_env = threshold + over/ratio. */
        int32_t gain = 256;
        if (comp_env > comp_threshold) {
            int32_t over   = comp_env - comp_threshold;
            int32_t target = comp_threshold + over / COMP_RATIO;
            gain = (target * 256) / comp_env;
        }

        /* Apply gain + makeup, then brickwall. */
        int32_t yl = (l * gain) >> 8;
        int32_t yr = (r * gain) >> 8;
        yl = (yl * COMP_MAKEUP_GAIN) >> 8;
        yr = (yr * COMP_MAKEUP_GAIN) >> 8;
        if (yl >  LIMIT_CEILING) yl =  LIMIT_CEILING;
        if (yl < -LIMIT_CEILING) yl = -LIMIT_CEILING;
        if (yr >  LIMIT_CEILING) yr =  LIMIT_CEILING;
        if (yr < -LIMIT_CEILING) yr = -LIMIT_CEILING;

        buf[2 * i]     = sat16(yl);
        buf[2 * i + 1] = sat16(yr);
    }
}

void compressor_adjust_threshold(int delta) {
    int v = (int)comp_threshold + delta;
    if (v < COMP_THRESHOLD_MIN) v = COMP_THRESHOLD_MIN;
    if (v > COMP_THRESHOLD_MAX) v = COMP_THRESHOLD_MAX;
    comp_threshold = v;
}

uint16_t compressor_get_threshold(void) {
    return (uint16_t)comp_threshold;
}

/* ---- one-shot init ------------------------------------------- */

void effects_init(void) {
    delay_init();
    reverb_init();
    comp_env       = 0;
    comp_threshold = COMP_THRESHOLD_DEFAULT;
}

File: zig/effects.zig
//! Master-bus stereo effects, ported from effects.c.
//!
//! effects.h is unchanged and remains the seam. Delay, Schroeder
//! reverb, soft saturation, compressor + brickwall limiter, and the
//! shared `sat16` helper that voice.c calls per sample.
//!
//! THE FIRST MODULE THAT CALLS ANOTHER
//!
//! Every prior port was a leaf. This one needs `arena_alloc`, so it
//! carries the `extern fn` half of the Principle V exception — the
//! declaration below is hand-matched against arena.h and nothing
//! checks it at compile time.
//!
//! It is a narrow exposure: one function, one argument, and a
//! mismatch cannot be subtle. `arena.h` declares
//! `void *arena_alloc(size_t n)`; a wrong width here would corrupt
//! every buffer allocated in effects_init(), which the bit-exact
//! golden catches on the first render.
//!
//! `arena_alloc` never returns null — arena.c treats OOM as a
//! programmer error and exits the process — so `.?` below matches the
//! C, which does not check either.
//!
//! Translation notes:
//!
//!   * Buffer pointers are `?[*]i16` initialised to null, matching the
//!     C statics that C zero-fills into .bss. `delay_init` /
//!     `reverb_init` test them for null to stay idempotent exactly as
//!     the C does.
//!
//!   * Every `int` in the header is `c_int` here, not `i32`: main.c's
//!     uniform flag table passes these directly.
//!
//!   * `soft_sat` widens to i64 before cubing, as the C does. |x| tops
//!     out at 32768 so x^3 reaches ~3.5e13 — it needs the 64-bit
//!     intermediate, and `>> 31` brings it back inside i32.
//!
//!   * Signed division uses @divTrunc (`over / COMP_RATIO`,
//!     `(target * 256) / comp_env`). Both numerators are non-negative
//!     at every reachable call site — `over` is guarded by
//!     `comp_env > comp_threshold` — so truncation and flooring agree
//!     today; @divTrunc is what C means regardless.
//!
//!   * All shifts are right shifts by constants. No `<<%` site.

/// From arena.h. See the module note on why this is hand-declared.
extern fn arena_alloc(n: usize) ?*anyopaque;

// ---- int16 saturating clamp -------------------------------------

export fn sat16(v: i32) i16 {
    if (v > 32767) return 32767;
    if (v < -32768) return -32768;
    return @intCast(v);
}

// ---- master-bus stereo delay ------------------------------------
// Two independent mono buffers, 250 ms at 48 kHz. Feed-forward +
// feedback: out = dry + tap*wet; buffer-write = dry + tap*feedback.

const DELAY_SAMPLES: usize = 12000;

var delay_l: ?[*]i16 = null;
var delay_r: ?[*]i16 = null;
var delay_idx: usize = 0;
/// 0..256, mix amount
var delay_wet: u16 = 100;
/// 0..200, capped to avoid runaway
var delay_feedback: u16 = 140;

fn allocZero(n_samples: usize) [*]i16 {
    const p = arena_alloc(n_samples * @sizeOf(i16));
    const buf: [*]i16 = @ptrCast(@alignCast(p.?));
    @memset(buf[0..n_samples], 0);
    return buf;
}

fn delayInit() void {
    // Idempotent: allocate on first call; afterwards clear in place.
    if (delay_l == null) {
        delay_l = allocZero(DELAY_SAMPLES);
        delay_r = allocZero(DELAY_SAMPLES);
    } else {
        @memset(delay_l.?[0..DELAY_SAMPLES], 0);
        @memset(delay_r.?[0..DELAY_SAMPLES], 0);
    }
    delay_idx = 0;
}

export fn delay_process(buf: [*]i16, frames: u32) void {
    const dl = delay_l.?;
    const dr = delay_r.?;
    var i: usize = 0;
    while (i < frames) : (i += 1) {
        const dry_l: i32 = buf[2 * i];
        const dry_r: i32 = buf[2 * i + 1];
        const tap_l: i32 = dl[delay_idx];
        const tap_r: i32 = dr[delay_idx];

        const out_l = dry_l + ((tap_l * @as(i32, delay_wet)) >> 8);
        const out_r = dry_r + ((tap_r * @as(i32, delay_wet)) >> 8);

        const fb_l = dry_l + ((tap_l * @as(i32, delay_feedback)) >> 8);
        const fb_r = dry_r + ((tap_r * @as(i32, delay_feedback)) >> 8);

        dl[delay_idx] = sat16(fb_l);
        dr[delay_idx] = sat16(fb_r);

        delay_idx += 1;
        if (delay_idx >= DELAY_SAMPLES) delay_idx = 0;

        buf[2 * i] = sat16(out_l);
        buf[2 * i + 1] = sat16(out_r);
    }
}

fn clampInt(v: c_int, lo: c_int, hi: c_int) c_int {
    if (v < lo) return lo;
    if (v > hi) return hi;
    return v;
}

export fn delay_adjust_wet(delta: c_int) void {
    delay_wet = @intCast(clampInt(@as(c_int, delay_wet) + delta, 0, 256));
}

export fn delay_adjust_feedback(delta: c_int) void {
    delay_feedback = @intCast(clampInt(@as(c_int, delay_feedback) + delta, 0, 200));
}

export fn delay_set_wet(v: c_int) void {
    delay_wet = @intCast(clampInt(v, 0, 256));
}

export fn delay_set_feedback(v: c_int) void {
    delay_feedback = @intCast(clampInt(v, 0, 200));
}

export fn delay_get_wet() u16 {
    return delay_wet;
}
export fn delay_get_feedback() u16 {
    return delay_feedback;
}

// ---- Schroeder reverb -------------------------------------------
// 4 parallel combs per channel summed into 2 series all-passes.
// Prime delays (Schroeder 1962, rescaled by 48000/44100) avoid
// metallic resonance; slightly different L/R delays keep the tail
// stereo.

const REV_C1L: u16 = 1693;
const REV_C2L: u16 = 1759;
const REV_C3L: u16 = 1621;
const REV_C4L: u16 = 1549;
const REV_C1R: u16 = 1721;
const REV_C2R: u16 = 1747;
const REV_C3R: u16 = 1613;
const REV_C4R: u16 = 1571;
const REV_AP1L: u16 = 241;
const REV_AP2L: u16 = 607;
const REV_AP1R: u16 = 251;
const REV_AP2R: u16 = 613;
/// ~0.70 in 8.8 fixed, RT60 ~1.5 s
const COMB_G: i32 = 180;
const AP_G: i32 = 180;

var rev_c1l: ?[*]i16 = null;
var rev_c2l: ?[*]i16 = null;
var rev_c3l: ?[*]i16 = null;
var rev_c4l: ?[*]i16 = null;
var rev_c1r: ?[*]i16 = null;
var rev_c2r: ?[*]i16 = null;
var rev_c3r: ?[*]i16 = null;
var rev_c4r: ?[*]i16 = null;
var rev_ap1l: ?[*]i16 = null;
var rev_ap2l: ?[*]i16 = null;
var rev_ap1r: ?[*]i16 = null;
var rev_ap2r: ?[*]i16 = null;

var i_c1l: u16 = 0;
var i_c2l: u16 = 0;
var i_c3l: u16 = 0;
var i_c4l: u16 = 0;
var i_c1r: u16 = 0;
var i_c2r: u16 = 0;
var i_c3r: u16 = 0;
var i_c4r: u16 = 0;
var i_ap1l: u16 = 0;
var i_ap2l: u16 = 0;
var i_ap1r: u16 = 0;
var i_ap2r: u16 = 0;

/// 0..256, mix amount
var reverb_wet: u16 = 60;
/// section-driven additive bias
var reverb_wet_bias: i8 = 0;

fn reverbInit() void {
    if (rev_c1l == null) {
        rev_c1l = allocZero(REV_C1L);
        rev_c1r = allocZero(REV_C1R);
        rev_c2l = allocZero(REV_C2L);
        rev_c2r = allocZero(REV_C2R);
        rev_c3l = allocZero(REV_C3L);
        rev_c3r = allocZero(REV_C3R);
        rev_c4l = allocZero(REV_C4L);
        rev_c4r = allocZero(REV_C4R);
        rev_ap1l = allocZero(REV_AP1L);
        rev_ap1r = allocZero(REV_AP1R);
        rev_ap2l = allocZero(REV_AP2L);
        rev_ap2r = allocZero(REV_AP2R);
    } else {
        @memset(rev_c1l.?[0..REV_C1L], 0);
        @memset(rev_c1r.?[0..REV_C1R], 0);
        @memset(rev_c2l.?[0..REV_C2L], 0);
        @memset(rev_c2r.?[0..REV_C2R], 0);
        @memset(rev_c3l.?[0..REV_C3L], 0);
        @memset(rev_c3r.?[0..REV_C3R], 0);
        @memset(rev_c4l.?[0..REV_C4L], 0);
        @memset(rev_c4r.?[0..REV_C4R], 0);
        @memset(rev_ap1l.?[0..REV_AP1L], 0);
        @memset(rev_ap1r.?[0..REV_AP1R], 0);
        @memset(rev_ap2l.?[0..REV_AP2L], 0);
        @memset(rev_ap2r.?[0..REV_AP2R], 0);
    }
    i_c1l = 0;
    i_c2l = 0;
    i_c3l = 0;
    i_c4l = 0;
    i_c1r = 0;
    i_c2r = 0;
    i_c3r = 0;
    i_c4r = 0;
    i_ap1l = 0;
    i_ap2l = 0;
    i_ap1r = 0;
    i_ap2r = 0;
}

/// y[n] = x[n-D] + g*y[n-D], delay line as a recirculating buffer.
inline fn combStep(buf: [*]i16, size: u16, idx: *u16, in: i16) i16 {
    const tap: i32 = buf[idx.*];
    const w = @as(i32, in) + ((tap * COMB_G) >> 8);
    buf[idx.*] = sat16(w);
    idx.* = (idx.* + 1) % size;
    return @intCast(tap);
}

/// All-pass: flat magnitude, shifts phase. Smooths the dense comb
/// output into a continuous tail.
inline fn apStep(buf: [*]i16, size: u16, idx: *u16, in: i16) i16 {
    const tap: i32 = buf[idx.*];
    const y = tap - ((@as(i32, in) * AP_G) >> 8);
    const w = @as(i32, in) + ((tap * AP_G) >> 8);
    buf[idx.*] = sat16(w);
    idx.* = (idx.* + 1) % size;
    return sat16(y);
}

export fn reverb_process(buf: [*]i16, frames: u32) void {
    var i: usize = 0;
    while (i < frames) : (i += 1) {
        const in_l: i16 = buf[2 * i];
        const in_r: i16 = buf[2 * i + 1];

        var sum_l: i32 = @as(i32, combStep(rev_c1l.?, REV_C1L, &i_c1l, in_l)) +
            combStep(rev_c2l.?, REV_C2L, &i_c2l, in_l) +
            combStep(rev_c3l.?, REV_C3L, &i_c3l, in_l) +
            combStep(rev_c4l.?, REV_C4L, &i_c4l, in_l);
        sum_l >>= 2;
        var sum_r: i32 = @as(i32, combStep(rev_c1r.?, REV_C1R, &i_c1r, in_r)) +
            combStep(rev_c2r.?, REV_C2R, &i_c2r, in_r) +
            combStep(rev_c3r.?, REV_C3R, &i_c3r, in_r) +
            combStep(rev_c4r.?, REV_C4R, &i_c4r, in_r);
        sum_r >>= 2;

        var ap_l = apStep(rev_ap1l.?, REV_AP1L, &i_ap1l, @intCast(sum_l));
        ap_l = apStep(rev_ap2l.?, REV_AP2L, &i_ap2l, ap_l);
        var ap_r = apStep(rev_ap1r.?, REV_AP1R, &i_ap1r, @intCast(sum_r));
        ap_r = apStep(rev_ap2r.?, REV_AP2R, &i_ap2r, ap_r);

        // Section bias on top of user wet, clamped to [0, 256].
        var eff_wet: i32 = @as(i32, reverb_wet) + @as(i32, reverb_wet_bias);
        if (eff_wet < 0) eff_wet = 0;
        if (eff_wet > 256) eff_wet = 256;
        const out_l = @as(i32, in_l) + ((@as(i32, ap_l) * eff_wet) >> 8);
        const out_r = @as(i32, in_r) + ((@as(i32, ap_r) * eff_wet) >> 8);
        buf[2 * i] = sat16(out_l);
        buf[2 * i + 1] = sat16(out_r);
    }
}

export fn reverb_adjust_wet(delta: c_int) void {
    reverb_wet = @intCast(clampInt(@as(c_int, reverb_wet) + delta, 0, 256));
}

export fn reverb_set_wet(v: c_int) void {
    reverb_wet = @intCast(clampInt(v, 0, 256));
}

export fn reverb_get_wet() u16 {
    return reverb_wet;
}

export fn reverb_set_wet_bias(bias: i8) void {
    reverb_wet_bias = bias;
}

// ---- soft saturation --------------------------------------------
// Cubic soft-clip: y = x - x^3 / 2^31. Linear for small x, smoothly
// compresses peaks. At full-scale input the output is ~50%; at
// typical levels (10-20% of full scale) the change is sub-1%.

inline fn softSat(x: i16) i16 {
    const xw: i64 = x;
    const x3: i64 = xw * xw * xw;
    const cubic: i32 = @intCast(x3 >> 31);
    return sat16(@as(i32, x) - cubic);
}

export fn saturate_process(buf: [*]i16, frames: u32) void {
    var i: usize = 0;
    while (i < frames) : (i += 1) {
        buf[2 * i] = softSat(buf[2 * i]);
        buf[2 * i + 1] = softSat(buf[2 * i + 1]);
    }
}

// ---- master compressor + brickwall limiter ----------------------
// Feed-forward, stereo-linked: the envelope tracks max(|L|,|R|) so
// gain reduction is identical on both channels and stereo imaging is
// preserved.

const COMP_THRESHOLD_DEFAULT: i32 = 20000;
const COMP_THRESHOLD_MIN: c_int = 8000;
const COMP_THRESHOLD_MAX: c_int = 30000;
/// 4:1 above threshold
const COMP_RATIO: i32 = 4;
/// (1-exp(-1/(0.005*48000)))*65536
const COMP_ATTACK_COEF: i32 = 268;
/// (1-exp(-1/(0.200*48000)))*65536
const COMP_RELEASE_COEF: i32 = 7;
/// ~+1 dB in 8.8 fixed
const COMP_MAKEUP_GAIN: i32 = 288;
/// brickwall, below int16 max so sat16 keeps margin
const LIMIT_CEILING: i32 = 32000;

var comp_env: i32 = 0;
var comp_threshold: i32 = COMP_THRESHOLD_DEFAULT;

export fn compressor_process(buf: [*]i16, frames: u32) void {
    var i: usize = 0;
    while (i < frames) : (i += 1) {
        const l: i32 = buf[2 * i];
        const r: i32 = buf[2 * i + 1];
        const al = if (l < 0) -l else l;
        const ar = if (r < 0) -r else r;
        const a = if (al > ar) al else ar;

        // Asymmetric envelope follower: fast attack, slow release.
        if (a > comp_env) {
            comp_env += ((a - comp_env) * COMP_ATTACK_COEF) >> 16;
        } else {
            comp_env -= ((comp_env - a) * COMP_RELEASE_COEF) >> 16;
        }

        // 1.0 below threshold, compressed above. gain is 8.8 fixed;
        // target_env = threshold + over/ratio.
        var gain: i32 = 256;
        if (comp_env > comp_threshold) {
            const over = comp_env - comp_threshold;
            const target = comp_threshold + @divTrunc(over, COMP_RATIO);
            gain = @divTrunc(target * 256, comp_env);
        }

        // Apply gain + makeup, then brickwall.
        var yl = (l * gain) >> 8;
        var yr = (r * gain) >> 8;
        yl = (yl * COMP_MAKEUP_GAIN) >> 8;
        yr = (yr * COMP_MAKEUP_GAIN) >> 8;
        if (yl > LIMIT_CEILING) yl = LIMIT_CEILING;
        if (yl < -LIMIT_CEILING) yl = -LIMIT_CEILING;
        if (yr > LIMIT_CEILING) yr = LIMIT_CEILING;
        if (yr < -LIMIT_CEILING) yr = -LIMIT_CEILING;

        buf[2 * i] = sat16(yl);
        buf[2 * i + 1] = sat16(yr);
    }
}

export fn compressor_adjust_threshold(delta: c_int) void {
    const v = clampInt(@as(c_int, comp_threshold) + delta, COMP_THRESHOLD_MIN, COMP_THRESHOLD_MAX);
    comp_threshold = @intCast(v);
}

export fn compressor_set_threshold(v: c_int) void {
    comp_threshold = @intCast(clampInt(v, COMP_THRESHOLD_MIN, COMP_THRESHOLD_MAX));
}

export fn compressor_get_threshold() u16 {
    return @intCast(comp_threshold);
}

// ---- one-shot init ----------------------------------------------

export fn effects_init() void {
    delayInit();
    reverbInit();
    comp_env = 0;
    comp_threshold = COMP_THRESHOLD_DEFAULT;
}
